File: test/BM_Index.js
// Patch to LaunchUp.js for creation of the search benchmakr database.

AE.LaunchUp.ComboBox = (function(self){
  // Private variables.
  var INPUT = null;
  var LIST = null;
  var DEFAULT_OVERLAY = null;
  var INPUT_TEXT = ''; // This variable keeps a copy of the original text that has been typed into the input field.
  var MIN_LENGTH = 0; // Minimum length of input before search starts.
  var SELECTED = -1; // Index of selected entry in dropdown list. (This is not a DATA id.)
  var SCHEDULER = new AE.Scheduler(500); // Allows to reduce too frequent search requests to 1 per 500ms.

  var old_initialize = self.initialize;
  self.initialize = function(input, list) {
    old_initialize(input, list);
    INPUT = input;
    LIST = list;

    // This sends ratings back to ruby as soon as you do a new search.
    var INPUTvalue = '';
    INPUT.onkeyup = function(event) {
      // For benchmarking the quality of the search algorithm(s).
      // Before we do a new search, we check if there are ratings of a previous search.
      //if (time_last_search && Number(new Date()) - time_last_search < 3) {
        var inputs = AE.$('input');
        // Continue only if there are search results.
        if (INPUT.value.length>0 && inputs.length >= 2) {
          var ratings = {};
          for (var i=0; i<inputs.length; i++) {
            var input = inputs[i];
            if (input.type === 'text' && input.id && input !== INPUT && input.value.length>0) {
              ratings[parseInt(input.id)] = parseInt(input.value);
            }
          }
          AE.Bridge.callRuby('optimization_database', [INPUTvalue, ratings]);
        }
      //}; time_last_search = Number(new Date());
      INPUTvalue = INPUT.value;

      // Original onkeyup:
      if (window.event) { var keycode = window.event.keyCode }
      else if (event) { var keycode = event.which };

      // arrow down: Select next entry.
      if (keycode == 40) {
        select(SELECTED + 1);
      }
      // arrow up: Select previous entry.
      else if (keycode == 38) {
        select(SELECTED - 1);
      }
      // enter key: Execute first entry.
      else if ( keycode == 13 && LIST.getElementsByTagName('li').length > 0 ) {
        if (SELECTED == -1) { select(0) };
        submit();
      }
      // other keys
      else {
        // To prevent complete searches at every key event, we use a scheduler to limit them to a minimum interval.
        SCHEDULER.replace(function() {
          SELECTED = -1;
          INPUT_TEXT = INPUT.value;
          if (INPUT_TEXT == '') {
            //emptyList(); // TODO
            AE.Style.show(DEFAULT_OVERLAY);
          } else if (INPUT_TEXT.replace(/^\s+|\s+$/g, '').length > MIN_LENGTH) {
            AE.Style.hide(DEFAULT_OVERLAY);
            AE.Bridge.callRuby('look_up', INPUT_TEXT, function(results){ rebuild(results) } );
          }
        });
      }
    }; // end function onkeyup
  };

  // This builds input elements into the list.
  // I could not easily insert the additional code, so this is a modified
  // copy of the original method:

  /* redefine: Build the suggestion list. */
  var rebuild = function(results) {
    if (!results) results = [];
    // Empty the results in the combo-list element.
    LIST.innerHTML = '';
    SELECTED = -1;
    // Generate an entry for each result.
    for (var i=0; i<results.length; i++) {
      // Container element(s)
      var entry = results[i];
      if (typeof(entry)==='undefined' || entry === null) { continue; }
      var div = document.createElement('div');
      var li = null;
      var li = document.createElement('li');
        li.setAttribute('id', 'entry_' + entry.id);
        li.setAttribute('title', entry.description||entry.name||'' );
        li.appendChild(div);
        li.entry = entry;
        // If the entry is enabled (or no such info known), its action can be executed.
        if (entry.enabled !== false) {
          li.onclick = function() {
            if (AE.debug) AE.Bridge.puts(Number(new Date())+' '+this.entry.id+' clicked'); // DEBUG
            submit(this.entry);
          }
        // Otherwise show it grayed.
        } else {
          li.setAttribute('class', 'grayed');
          li.onclick = function() {
            History.add(this.entry);
          }
        };
      // <INSERTION
      // For benchmarking the quality of the search algorithm(s).
      // Create a text input where I can insert a rating (0-10) for each result.
      li.onclick = null;
      var input = document.createElement('input');
      input.setAttribute('id', entry.id);
      input.style.maxWidth = '3em';
      input.style.float = 'right';
      div.appendChild(input);
      // INSERTION>
      // Icon for the command.
      if (typeof(entry.icon) !== 'undefined' && entry.icon !== null && entry.icon !== '') {
        var img = document.createElement('img');
        img.setAttribute('src', entry.icon);
        div.appendChild(img);
      };
      if (AE.debug) { // DEBUG
        // Score (for debugging correct order).
        div.appendChild(
          document.createElement('span').appendChild(
            document.createTextNode( Math.round(100*entry.score)/100 )));
        // ID (for debugging).
        div.appendChild(
          document.createElement('span').appendChild(
            document.createTextNode( '('+entry.id+')' )));
      };
      // Name of the command.
      var span = document.createElement('span');
      span.setAttribute('class', 'name');
      span.appendChild(
        document.createTextNode(entry.name||''));
      div.appendChild(span);
      // Category of the command.
      if (entry.category) {
        var span = document.createElement('span');
        span.setAttribute('class', 'category');
        span.appendChild(
          document.createTextNode(' '+entry.category));
        div.appendChild(span);
      }
      // Description of the command.
      if (AE.debug && typeof entry.dbg_info !== 'undefined' && entry.dbg_info !== null) { entry.description = entry.dbg_info + ' ' + entry.description } // DEBUG
      if (entry.description !== entry.name) {
        var span = document.createElement('span');
        span.setAttribute('class', 'description');
        span.appendChild(
          document.createTextNode(' '+entry.description||''));
        div.appendChild(span);
      }
      // Attach the entry into the DOM.
      LIST.appendChild(li);
    }
    AE.Dialog.adjustSize();
  }; // end function rebuild

  return self;
})(AE.LaunchUp.ComboBox)

File: test/BM_Index.rb
=begin
This file allows to benchmark the search algorithm of LaunchUp by comparing the
order of results with a database of expected results. It can automatically run
hundreds of searches and was supposed to show how a modification/optimization of
the search algorithm impacts the quality/relevance of results.

Load it after LaunchUp, but before opening the LaunchUp dialog. It adds input
fields into the search results where you can rate the placement/order of results
from 0 to 10 (best).
Run AE::LaunchUp.run_benchmark to get a score (around 170–200) of the current
search algorithm (heigher number = better).
=end



module AE



module LaunchUp



BM_INDEX_ROOT = File.dirname(__FILE__) unless defined?(BM_INDEX_ROOT)

#require 'sketchup.rb'
load(File.join(File.dirname(BM_INDEX_ROOT), 'src', 'ae_LaunchUp', 'LaunchUp.rb'))

# Data for optimizing the search algorithm.
load(File.join(BM_INDEX_ROOT, 'BM_Index_database.rb'))



def self.save_optimization_database
  file = File.join(BM_INDEX_ROOT, "BM_Index_database.rb")
  File.open(file, "w"){|f|
    f.puts("# This file contains search strings with expected results rated from 0 (bad) to 10 (good).\nmodule AE\nmodule LaunchUp\n@optimization_database=")
    f.print(@optimization_database.inspect)
    f.puts("\nend\nend")
  }
  return true
rescue
  return false
end



def self.run_benchmark
  load(File.join(File.dirname(BM_INDEX_ROOT), 'src', 'ae_LaunchUp', 'Index.rb'))
  rating = 0
  t = Time.now.to_f
  @optimization_database.each{|search_word, hash|
    results = self.look_up(search_word, 10)
    results.each_with_index{|result, i|
      id = result[:id]
      next unless hash.include?(id) && hash[id].is_a?(Numeric)
      rating += (10-i) * hash[id]
    }
  }
  puts("Time: #{(Time.now.to_f - t)/(@optimization_database.length|1)}")
  rating /= @optimization_database.length.to_f
  return rating
end



unless file_loaded?(File.basename(__FILE__))

  class << self
    alias_method(:show_dialog_old, :show_dialog)
  end

end # end unless



  # Changes to the LaunchUp dialog that add input fields next to each search result.
  # Instead of executing an action, you can enter a rating of each result from 0 to 10.
  def self.show_dialog
    self.show_dialog_old
    @launchdlg.add_action_callback("initialize") {|dlg, param|
      TRANSLATE.webdialog(dlg)
      # Prevent that the dialog collapses:
      dlg.execute_script("AE.Dialog.addOnFocus = function(){}; AE.Dialog.addOnBlur = function(){};")
      # Inject code into the ComboBox:
      js_file = File.expand_path(File.join(BM_INDEX_ROOT, "BM_Index.js"))
      dlg.execute_script("
      var s = document.createElement('script');
      s.setAttribute('src','#{js_file}');
      /*s.onload = function(){
      };*/
      document.body.appendChild(s);
      ")
      UI.start_timer(1, false){dlg.execute_script("AE.LaunchUp.initialize(#{@options.get_json});")}
    }

    # Collect data for the benchmark database.
    @launchdlg.add_action_callback("optimization_database") { |dlg, param|
      search_word, hash = eval(param) rescue next
      next if hash.empty?
      search_word.downcase!
      @optimization_database ||= {}
      @optimization_database[search_word] ||= {}
      hash.each{|id, rating|
        id = id.to_s.to_i unless id.is_a?(Fixnum)
        @optimization_database[search_word][id] ||= rating
        old_rating = @optimization_database[search_word][id]
        @optimization_database[search_word][id] = (2*old_rating + rating) / 3.0
      }
      @optimization_database[search_word].merge!(hash)
    }
  end # def



end # module LaunchUp



end # module AE
